docs/internal/generics.md

```diff
@@ -156,3 +156,110 @@ let z : U32 & imm = obj.convert(y); // ERROR: Not able to define what type y is
 
 ## Specialisation
 
+Type parameters make classes generic to share implementation for a range of different types ([parametric polymorphism](https://en.wikipedia.org/wiki/Parametric_polymorphism)).
+But concrete classes need monomorphic types to be defined at compile time, so that we can have a concrete implementation of a parametric type.
+This conversion is called [reification](https://en.wikipedia.org/wiki/Reification_(computer_science)) and is done via a reachability analysis of all possible definitions of each polymorphic type.
+
+Example:
+```ts
+class Generic[A] { field : A & imm; ... };
+...
+// Here, `A` is `U32`
+let foo : Generic[U32] = ...;
+
+// Here, `A` is `MyClass`
+let bar : Generic[MyClass] = ...;
+```
+
+In the example above, there is one polymorphic class, `Generic[A]` where `A` is the polymorphic type parameter, and two concrete classes (specialisations): `Generic[U32]` and `Generic[MyClass]`.
+
+The compiler generates the required specialisations and replaces each definition with its respective concrete classes, then removes the generic class.
+Since generic classes can only be used by specialisation, all other (potentially infinite) possibilities of types parameters don't exist in the program, so the generic class is elided and we're left with only the specialisations.
+
+The end result would be:
+```ts
+// Definition of monomorphic classes only
+class Generic_U32 { field : U32 & imm; ... };
+class Generic_MyClass { field : MyClass & imm; ... }
+
+// Direct use of the concrete classes
+let foo : Generic_U32 = ...;
+let bar : Generic_MyClass = ...;
+```
+
+Note that type unions (`A | B`) are still monomorphic types.
+To resolve which implementation works with each inner type (`A` or `B`) one must use `match` inside the method.
+
+Example:
+```ts
+class Generic[A] {
+  field : A & imm;
+  // Implement constructor with match on types
+  add(other : A & imm) {
+    match (A)
+    { Numeric => field + other }
+    { String => field.append(other) }
+    { _ => Throw[None] } // Or some better error handling
+  }
+}
+...
+// This creates a monomorphic union type
+let foo : Generic[U32 | String] = ...;
+
+// This creates a monomorphic single type
+let bar : Generic[U32] = ...;
+
+// This union of Numeric types isn't quite static
+let baz : Generic[U32 | F64] = ...;
+```
+
+On the first case, `foo` is a monomorphic union between `U32` and `String`, which means, at run-time, either can be passed to `add` and we can't decide at compile time.
```

Comment:
'monomorphic union' feels like an oxymoron to me, but that might just be me.

Reply:
Now, it does to me too, but I wanted to be clear. It is not obvious to people that haven't been looking at type systems for the past few years, as it certainly was't clear to me when I first started working on Verona.

Reply:
put between parenthesis to make it clear that it's redundant.